draft-ietf-ppm-dap.md

```diff
@@ -575,29 +560,36 @@ particularly important for controlling the amount of noise added to reports
 by Clients (or added to aggregate shares by Aggregators) in order to achieve
 differential privacy.
 
-To get the aggregate of the next batch, the Collector issues a query of type
-`next-batch`. (See FixedSizeQuery in {{query}}.) The response to this query (see
-{{collect-flow}}) includes a "batch ID", which can be used to query the same
-batch later on. This is done using the `by-batch-id` batch-query type.
-
 For this query type, the Aggregators group reports into arbitrary batches such
-that each batch has roughly the same number of reports. In addition to the
-minumum batch size common to all query types, the conifugration includes a
-"maximum batch size", `max_batch_size`, that determines maximum number of
-reports per batch.
+that each batch has roughly the same number of reports. These batches are
+identified by "batch IDs", which are numerical identifiers allocated
+consecutively starting from 0.
```

Comment:
Should we say "monotonic" to allow discontinuities in batch IDs? Does anything bad happen if there are discontinuities in batch IDs? I assume we already have text instructing aggregators what to do if they don't recognize a batch ID.

Reply:
I think we don't want discontinuities: the idea of this approach is that the Collector can send requests for batch 0, followed by batch 1, followed by batch 2, and so on -- i.e. that the Collector can predict the sequence of batch identifiers. While a failed aggregation or technical issues might lead to an uncollectable batch, which (hopefully) is communicated well enough for the Collector to realize it won't be able to collect that batch, allowing the Leader to skip arbitrary batch IDs would currently be painful[1]. For example, if the Leader decided to skip `10^6` batch IDs, I suppose the Collector would have to walk through `10^6` collect requests getting an "unknown batch" before finding the next "good" batch.


[1] I think the best way to reduce this pain, if we decide we need to, would be to have some way for the Collector to ask the Leader which batch IDs are usable/acceptable. That's what I was trying to get at with the now-removed `current-batch` functionality, to allow Collectors to startup against Aggregators that have been running for awhile without the need for out-of-band communication. Like with that functionality, IMO we should see whether we need to implement something more complicated before we specify it.

Reply:
> LGTM. One potential downside is that the Leader can no longer choose batch IDs at random, since it needs to fill batch IDs consecutively. I wonder if this could potentially create a processing bottleneck for the Leader. (I don't know the answer so I'm hoping other reviewers will take the time to think this through!)

I'm not _too_ worried about this over-and-above the current design: I think there is already effectively a "serialization point" required when the Leader assigns the next aggregation job to a batch -- at that point, it needs to know how many outstanding aggregation jobs are assigned to batches, and how many reports are in those aggregation jobs, in order to decide the batch ID to use for its next aggregation job. I think it needs to know this since, if this knowledge is unavailable or imprecise, it might assign too many reports to a single batch, leading to those reports being uncollectable.

Once you have a serialization point at point-of-generation-of-aggregation-job, choosing a new batch ID would be something like the pseudo-SQL `SELECT MAX(batch_id) + 1 FROM batches WHERE batch_full = false`.

That said -- maybe the existing serialization point will be too much of a performance hit, with or without generating a consecutive batch ID? I guess implementation experience will tell. I'm not too worried since this serialization point wouldn't affect the aggregation protocol, which is likely to be the costliest part of the overall protocol, but it's hard to be certain without actually implementing.

(also, did I miss something with the above? maybe there's some kinds of imprecision in knowledge of assignment of client reports/aggregation jobs to batches that would allow a serialization point to be safely skipped?)

Reply:
OK, then do we need a MUST describing how the leader should assign batch IDs?

Reply:
Sure -- the existing text in the "Fixed-size Queries" section defined them this way, but it's probably valuable to reiterate this point in the "Leader Initialization" section since that section defines the specific protocol steps. I did so in terms of a `MUST`.